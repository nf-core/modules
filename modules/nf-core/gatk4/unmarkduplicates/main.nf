process GATK4_UNMARKDUPLICATES {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/gatk4:4.6.1.0--py310hdfd78af_0':
        'biocontainers/gatk4:4.6.1.0--py310hdfd78af_0' }"

    input:
    tuple val(meta), path(bam)

    output:
    tuple val(meta), path("${prefix}.bam"),     emit: bam
    tuple val(meta), path("${prefix}.bai"),     emit: bai
    path "versions.yml",                emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}_UnmarkDuplicates"
    def input_list = bam.collect{"--input $it"}.join(' ')
    def avail_mem = 3072
    if (!task.memory) {
        log.info '[GATK UnmarkDuplicates] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this.'
    } else {
        avail_mem = (task.memory.mega*0.8).intValue()
    }

    """
    gatk --java-options "-Xmx${avail_mem}M -XX:-UsePerfData" \\
        UnmarkDuplicates \\
        $input_list \\
        --output ${prefix}.bam \\
        $args

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        gatk4: \$(echo \$(gatk --version 2>&1) | sed 's/^.*(GATK) v//; s/ .*\$//')
    END_VERSIONS
    """

    stub:
    prefix = task.ext.prefix ?: "${meta.id}_UnmarkDuplicates"
    prefix_no_suffix = task.ext.prefix ? prefix.tokenize('.')[0] : "${meta.id}"
    """
    touch ${prefix}.bam
    touch ${prefix}.bai

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        gatk4: \$(echo \$(gatk --version 2>&1) | sed 's/^.*(GATK) v//; s/ .*\$//')
    END_VERSIONS
    """
}
