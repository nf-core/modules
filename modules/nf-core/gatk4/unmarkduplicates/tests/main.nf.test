nextflow_process {

    name "Test Process GATK4_UNMARKDUPLICATES"
    script "../main.nf"
    process "GATK4_UNMARKDUPLICATES"

    tag "modules"
    tag "modules_nfcore"
    tag "gatk4"
    tag "gatk4/unmarkduplicates"
    tag "gatk4/markduplicates"


    test("sarscov2 - bam") {

        setup {
            run("GATK4_MARKDUPLICATES") {
                script "../../../gatk4/markduplicates/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test', single_end:false ],
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                    ]
                    input[1] = []
                    input[2] = []
                    """
                }
            }
        }

        when {
            process {
                """
                input[0] = GATK4_MARKDUPLICATES.out.bam.groupTuple(by:[0])
                """
            }
        }


        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("homo_sapiens - multiple bam") {

        setup {
            run("GATK4_MARKDUPLICATES") {
                script "../../../gatk4/markduplicates/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test', single_end:false ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                            file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test2.paired_end.sorted.bam', checkIfExists: true)
                        ]
                    ]
                    input[1] = []
                    input[2] = []
                """
                }
            }
        }

        when {
            process {
                """
                input[0] = GATK4_MARKDUPLICATES.out.bam.groupTuple(by:[0])
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.bam).match("multi bam") },
                { assert snapshot(process.out.bai).match("multi bai") },
                { assert snapshot(process.out.versions).match("multi versions") },
            )
        }

    }

    test("sarscov2 - bam - stub") {

        options "-stub"

        setup {
            run("GATK4_MARKDUPLICATES") {
                script "../../../gatk4/markduplicates/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test', single_end:false ],
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                    ]
                    input[1] = []
                    input[2] = []
                    """
                }
            }
        }

        when {
            process {
                """
                input[0] = GATK4_MARKDUPLICATES.out.bam.groupTuple(by:[0])
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
