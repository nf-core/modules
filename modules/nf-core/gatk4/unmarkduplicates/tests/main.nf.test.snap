{
    "multi bam": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test_UnmarkDuplicates.bam:md5,3723a12f43c15628a2dde0e68e1ae343"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-28T16:00:53.693147"
    },
    "sarscov2 - bam - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_UnmarkDuplicates.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_UnmarkDuplicates.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,38963b2a5709c7b6f3d97915cc8bfe50"
                ],
                "bai": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_UnmarkDuplicates.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "bam": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_UnmarkDuplicates.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,38963b2a5709c7b6f3d97915cc8bfe50"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-28T16:01:20.128405"
    },
    "multi bai": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test_UnmarkDuplicates.bai:md5,033bea816cfb968ae01b5451e9ac2ac4"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-28T16:00:53.702211"
    },
    "sarscov2 - bam": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_UnmarkDuplicates.bam:md5,c2728362b7cbd63767bd0bc0fbf26e16"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_UnmarkDuplicates.bai:md5,010fbec4550a52cc16c4b719b3b358a1"
                    ]
                ],
                "2": [
                    "versions.yml:md5,38963b2a5709c7b6f3d97915cc8bfe50"
                ],
                "bai": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_UnmarkDuplicates.bai:md5,010fbec4550a52cc16c4b719b3b358a1"
                    ]
                ],
                "bam": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_UnmarkDuplicates.bam:md5,c2728362b7cbd63767bd0bc0fbf26e16"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,38963b2a5709c7b6f3d97915cc8bfe50"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-28T15:59:57.655713"
    },
    "multi versions": {
        "content": [
            [
                "versions.yml:md5,38963b2a5709c7b6f3d97915cc8bfe50"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-28T16:00:53.70895"
    }
}